File: mcp/src/adb-link.ts
import { adbReady, adbShell, adbShellText, shellQuote } from "./adb.js";
import type { Link } from "./link.js";

/**
 * Controls the device purely through `adb`, no on-device app required.
 *
 * This is the quick-connect path: enable USB debugging (or adb-over-Wi-Fi),
 * plug in, and a model can drive the phone. Every protocol op is answered by
 * shelling out — `input` for gestures and keys, `uiautomator dump` for the tree,
 * `screencap` for pixels — and shaped into the same response the on-device app
 * would return, so the tool layer never learns which backend it is talking to.
 *
 * What adb cannot honestly do, it says so about rather than faking. There is no
 * per-app grant model here: adb holds shell-level authority over the whole
 * device, and that is disclosed in the capability report and the session notice
 * rather than dressed up as something narrower.
 */
export class AdbLink implements Link {
  /** Centres of the nodes from the last ui_tree, so taps can address by id. */
  private nodeCenters = new Map<string, { x: number; y: number }>();
  private nodeBounds = new Map<string, Bounds>();

  close(): void {
    // Nothing persistent to close; every call spawns adb afresh.
  }

  async send(request: any, _timeoutMs = 30_000): Promise<any> {
    try {
      switch (request?.op) {
        case "capabilities":
          return await this.capabilities();
        case "session_begin":
          return await this.sessionBegin(request.ttlSec ?? 300);
        case "session_end":
          return { ok: true, foregroundChanged: false };
        case "grants_list":
          // adb has no per-app grant model; the whole device is reachable.
          return { ok: true, grants: [] };
        case "apps_list":
          return await this.appsList(request.grantedOnly ?? true);
        case "ui_tree":
          return await this.uiTree();
        case "screenshot":
          return await this.screenshot();
        case "notifications":
          return await this.notifications();
        case "tap":
          return await this.tap(request);
        case "long_press":
          return await this.longPress(request);
        case "swipe":
          return await this.swipe(request);
        case "scroll":
          return await this.scroll(request);
        case "type":
          return await this.type(request);
        case "key":
          return await this.key(request.key);
        case "launch":
          return await this.launch(request.pkg);
        case "shell":
          return await this.shell(request.cmd);
        case "clipboard_get":
        case "clipboard_set":
          return {
            ok: false,
            error:
              "clipboard is not available over adb — pair the on-device app for clipboard access",
          };
        default:
          return { ok: false, error: `unsupported op: ${request?.op}` };
      }
    } catch (e) {
      const message = e instanceof Error ? e.message : String(e);
      // ENOENT means adb itself is missing; make that actionable.
      if (message.includes("ENOENT")) {
        return {
          ok: false,
          error:
            "adb not found on PATH — install platform-tools, or set MCPSERVED_ADB to the adb binary",
        };
      }
      return { ok: false, error: message };
    }
  }

  // ---- capabilities & session -------------------------------------------

  private async capabilities(): Promise<any> {
    if (!(await adbReady())) {
      return {
        ok: false,
        error:
          "no adb device — attach over USB and run `adb devices`, or `adb connect <ip>:5555` for Wi-Fi",
      };
    }

    let root = false;
    try {
      const id = await adbShellText("su -c id", { timeoutMs: 5_000 });
      root = /uid=0/.test(id);
    } catch {
      root = false;
    }

    // adb shell is itself a shell-level backend; report it as such so the shell
    // tool is offered. "shizuku" here stands for that ADB-level authority.
    const caps = [
      "TREE",
      "GESTURE",
      "TEXT_INPUT",
      "GLOBAL_KEYS",
      "CAPTURE_SILENT",
      "NOTIFICATIONS",
      "SHELL_SHIZUKU",
    ];
    if (root) caps.push("SHELL_ROOT");

    return { ok: true, caps, root, shizuku: true, a11y: false };
  }

  private async sessionBegin(ttlSec: number): Promise<any> {
    // Best-effort wake so the screen is on and unlocked enough to act on.
    try {
      await adbShell("input keyevent 224", { timeoutMs: 5_000 }); // WAKEUP
      await adbShell("wm dismiss-keyguard", { timeoutMs: 5_000 });
    } catch {
      // A device that will not wake is still worth trying to drive.
    }
    return {
      ok: true,
      sessionId: "adb",
      expiresAtEpochMs: Date.now() + ttlSec * 1000,
      foregroundChanged: false,
    };
  }

  // ---- reading ----------------------------------------------------------

  private async appsList(grantedOnly: boolean): Promise<any> {
    // No grant model over adb, so "granted" is universal. When grantedOnly is
    // set, narrow to third-party packages to keep the list to what a task
    // usually means rather than dumping every system component.
    const flag = grantedOnly ? " -3" : "";
    const out = await adbShellText(`pm list packages${flag}`);
    const apps = out
      .split("\n")
      .map((l) => l.trim())
      .filter((l) => l.startsWith("package:"))
      .map((l) => l.slice("package:".length))
      .filter(Boolean)
      .map((pkg) => ({ pkg, label: pkg, granted: true }));
    return { ok: true, apps };
  }

  private async uiTree(): Promise<any> {
    // uiautomator writes to a file; read it back byte-exact via exec-out.
    let dumpPath = "/sdcard/window_dump.xml";
    try {
      const out = await adbShellText("uiautomator dump", { timeoutMs: 15_000 });
      const m = out.match(/dumped to:\s*(\S+)/);
      if (m) dumpPath = m[1];
    } catch (e) {
      return { ok: false, error: `uiautomator dump failed: ${(e as Error).message}` };
    }

    const xmlRes = await adbShell(`cat ${dumpPath}`, { binary: true, timeoutMs: 15_000 });
    if (xmlRes.code !== 0) {
      return { ok: false, error: "could not read the ui dump" };
    }
    const xml = xmlRes.stdout.toString("utf8");

    const parsed = parseUiAutomator(xml);
    this.nodeCenters.clear();
    this.nodeBounds.clear();
    for (const n of parsed.nodes) {
      this.nodeCenters.set(n.id, {
        x: Math.round((n.bounds.l + n.bounds.r) / 2),
        y: Math.round((n.bounds.t + n.bounds.b) / 2),
      });
      this.nodeBounds.set(n.id, n.bounds);
    }

    return {
      ok: true,
      pkg: parsed.pkg,
      activity: null,
      nodes: parsed.nodes,
      pruned: parsed.pruned,
      foregroundChanged: false,
    };
  }

  private async screenshot(): Promise<any> {
    const res = await adbShell("screencap -p", { binary: true, timeoutMs: 20_000 });
    if (res.code !== 0 || res.stdout.length === 0) {
      return { ok: false, error: "screencap failed" };
    }
    const png = res.stdout;
    const { w, h } = pngSize(png);
    // maxPx is not honoured: adb has no image scaler on the host side. The raw
    // frame is returned; the tool description already steers toward ui_tree.
    return {
      ok: true,
      mime: "image/png",
      b64: png.toString("base64"),
      w,
      h,
      foregroundChanged: false,
    };
  }

  private async notifications(): Promise<any> {
    // Best-effort parse of the notification dump. adb exposes no clean list, so
    // this is heuristic; the on-device app's NotificationListener is the honest
    // path and is offered as the upgrade.
    let out = "";
    try {
      out = await adbShellText("dumpsys notification --noredact", { timeoutMs: 10_000 });
    } catch {
      return { ok: true, items: [] };
    }

    const items: Array<{ pkg: string; key: string; title: string | null; text: string | null; postedAtEpochMs: number }> = [];
    const blocks = out.split(/NotificationRecord\(/).slice(1);
    for (const block of blocks) {
      const pkg = block.match(/pkg=(\S+)/)?.[1] ?? "";
      const title = block.match(/android\.title=String \(([^)]*)\)/)?.[1] ?? null;
      const text = block.match(/android\.text=String \(([^)]*)\)/)?.[1] ?? null;
      if (!pkg) continue;
      items.push({ pkg, key: pkg, title, text, postedAtEpochMs: 0 });
    }
    return { ok: true, items };
  }

  // ---- acting -----------------------------------------------------------

  private resolvePoint(req: { nodeId?: string; x?: number; y?: number }): { x: number; y: number } | null {
    if (req.nodeId) return this.nodeCenters.get(req.nodeId) ?? null;
    if (typeof req.x === "number" && typeof req.y === "number") return { x: req.x, y: req.y };
    return null;
  }

  private async tap(req: any): Promise<any> {
    const p = this.resolvePoint(req);
    if (!p) return { ok: false, error: "no node id (call ui_tree first) or x,y given" };
    await adbShellText(`input tap ${p.x} ${p.y}`);
    return { ok: true, foregroundChanged: false };
  }

  private async longPress(req: any): Promise<any> {
    const p = this.resolvePoint(req);
    if (!p) return { ok: false, error: "no node id or x,y given" };
    const ms = req.ms ?? 500;
    await adbShellText(`input swipe ${p.x} ${p.y} ${p.x} ${p.y} ${ms}`);
    return { ok: true, foregroundChanged: false };
  }

  private async swipe(req: any): Promise<any> {
    const ms = req.ms ?? 300;
    await adbShellText(`input swipe ${req.x1} ${req.y1} ${req.x2} ${req.y2} ${ms}`);
    return { ok: true, foregroundChanged: false };
  }

  private async scroll(req: any): Promise<any> {
    const b = req.nodeId ? this.nodeBounds.get(req.nodeId) : undefined;
    if (!b) return { ok: false, error: "unknown node id — call ui_tree first" };
    const cx = Math.round((b.l + b.r) / 2);
    const cy = Math.round((b.t + b.b) / 2);
    const w = b.r - b.l;
    const h = b.b - b.t;
    // A swipe opposite the reading direction of travel: scrolling DOWN reveals
    // content below, which is a swipe upward.
    let x1 = cx, y1 = cy, x2 = cx, y2 = cy;
    switch (req.dir) {
      case "DOWN": y1 = b.t + Math.round(h * 0.7); y2 = b.t + Math.round(h * 0.3); break;
      case "UP": y1 = b.t + Math.round(h * 0.3); y2 = b.t + Math.round(h * 0.7); break;
      case "RIGHT": x1 = b.l + Math.round(w * 0.7); x2 = b.l + Math.round(w * 0.3); break;
      case "LEFT": x1 = b.l + Math.round(w * 0.3); x2 = b.l + Math.round(w * 0.7); break;
      default: return { ok: false, error: `unknown direction: ${req.dir}` };
    }
    await adbShellText(`input swipe ${x1} ${y1} ${x2} ${y2} 300`);
    return { ok: true, foregroundChanged: false };
  }

  private async type(req: any): Promise<any> {
    if (req.nodeId) {
      const p = this.nodeCenters.get(req.nodeId);
      if (p) await adbShellText(`input tap ${p.x} ${p.y}`);
    }
    // `input text` takes spaces as %s and cannot express newlines or most
    // non-ASCII. Good enough for field entry; the app path handles the rest.
    const encoded = String(req.text ?? "").replace(/ /g, "%s");
    await adbShellText(`input text ${shellQuote(encoded)}`);
    return { ok: true, foregroundChanged: false };
  }

  private async key(key: string): Promise<any> {
    const map: Record<string, string> = {
      BACK: "4",
      HOME: "3",
      RECENTS: "187",
      ENTER: "66",
      DELETE: "67",
    };
    if (key === "NOTIFICATIONS") {
      await adbShellText("cmd statusbar expand-notifications");
      return { ok: true, foregroundChanged: false };
    }
    const code = map[key];
    if (!code) return { ok: false, error: `unknown key: ${key}` };
    await adbShellText(`input keyevent ${code}`);
    return { ok: true, foregroundChanged: false };
  }

  private async launch(pkg: string): Promise<any> {
    const r = await adbShell(
      `monkey -p ${shellQuote(pkg)} -c android.intent.category.LAUNCHER 1`,
      { timeoutMs: 10_000 },
    );
    if (r.code !== 0) {
      return { ok: false, error: `could not launch ${pkg}` };
    }
    return { ok: true, foregroundChanged: true };
  }

  private async shell(cmd: string): Promise<any> {
    const r = await adbShell(cmd, { timeoutMs: 30_000 });
    const text = (r.stdout.toString("utf8") + (r.stderr ? `\n${r.stderr}` : "")).trim();
    return { ok: r.code === 0, error: r.code === 0 ? null : text || `exited ${r.code}`, text };
  }
}

// ---- parsing helpers ------------------------------------------------------

interface Bounds {
  l: number;
  t: number;
  r: number;
  b: number;
}

interface AdbNode {
  id: string;
  cls: string;
  bounds: Bounds;
  depth: number;
  text?: string;
  desc?: string;
  clickable: boolean;
  editable: boolean;
  scrollable: boolean;
  checked?: boolean;
  enabled: boolean;
}

function decodeEntities(s: string): string {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&apos;/g, "'");
}

function parseBounds(s: string | undefined): Bounds | null {
  const m = s?.match(/\[(-?\d+),(-?\d+)\]\[(-?\d+),(-?\d+)\]/);
  if (!m) return null;
  return { l: +m[1], t: +m[2], r: +m[3], b: +m[4] };
}

/**
 * Parses a uiautomator hierarchy dump into the flat node list the tools expect.
 *
 * Pure layout containers — nodes with no text, no content description, and no
 * interactive flag — are dropped and counted, matching what the on-device
 * Pruner does, so the tree the model reads is interaction, not scaffolding.
 */
function parseUiAutomator(xml: string): { pkg: string; nodes: AdbNode[]; pruned: number } {
  const nodes: AdbNode[] = [];
  let pruned = 0;
  let pkg = "";
  let depth = 0;
  let seq = 0;

  const tag = /<node\b([^>]*?)(\/?)>|<\/node>/g;
  let m: RegExpExecArray | null;

  while ((m = tag.exec(xml)) !== null) {
    if (m[0] === "</node>") {
      depth = Math.max(0, depth - 1);
      continue;
    }

    const attrsRaw = m[1];
    const selfClosing = m[2] === "/";
    const attrs: Record<string, string> = {};
    const attrRe = /([\w-]+)="([^"]*)"/g;
    let a: RegExpExecArray | null;
    while ((a = attrRe.exec(attrsRaw)) !== null) {
      attrs[a[1]] = decodeEntities(a[2]);
    }

    const bounds = parseBounds(attrs["bounds"]);
    const thisDepth = depth;
    if (!selfClosing) depth += 1;
    if (!bounds) continue;
    if (!pkg && attrs["package"]) pkg = attrs["package"];

    const text = attrs["text"] || undefined;
    const desc = attrs["content-desc"] || undefined;
    const clickable = attrs["clickable"] === "true";
    const longClickable = attrs["long-clickable"] === "true";
    const scrollable = attrs["scrollable"] === "true";
    const checkable = attrs["checkable"] === "true";
    const cls = (attrs["class"] || "").split(".").pop() || attrs["class"] || "View";
    const editable = /EditText/i.test(attrs["class"] || "");

    const interactive = clickable || longClickable || scrollable || checkable || editable;
    if (!interactive && !text && !desc) {
      pruned += 1;
      continue;
    }
    if (bounds.r - bounds.l <= 0 || bounds.b - bounds.t <= 0) {
      pruned += 1;
      continue;
    }

    const rid = attrs["resource-id"];
    const idBase = rid ? rid.split("/").pop() || rid : cls.toLowerCase();
    const id = `${idBase}#${seq++}`;

    nodes.push({
      id,
      cls,
      bounds,
      depth: thisDepth,
      text,
      desc,
      clickable,
      editable,
      scrollable,
      checked: checkable ? attrs["checked"] === "true" : undefined,
      enabled: attrs["enabled"] !== "false",
    });
  }

  return { pkg, nodes, pruned };
}

/** Reads width and height from a PNG's IHDR chunk. */
function pngSize(png: Buffer): { w: number; h: number } {
  // IHDR width/height are the two big-endian uint32s at offset 16.
  if (png.length >= 24 && png.toString("ascii", 12, 16) === "IHDR") {
    return { w: png.readUInt32BE(16), h: png.readUInt32BE(20) };
  }
  return { w: 0, h: 0 };
}

File: desktop/src/main/kotlin/com/hereliesaz/mcpserved/desktop/net/AdbLink.kt
package com.hereliesaz.mcpserved.desktop.net

import com.hereliesaz.mcpserved.desktop.adb.Adb
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import java.util.Base64

/**
 * Controls the device purely through `adb`, no on-device app required.
 *
 * This is the quick-connect path: enable USB debugging (or adb-over-Wi-Fi), plug
 * in, and a model can drive the phone. Every protocol op is answered by shelling
 * out — `input` for gestures and keys, `uiautomator dump` for the tree,
 * `screencap` for pixels — and shaped into the same response the on-device app
 * would return, so the tool layer never learns which backend it is talking to.
 *
 * What adb cannot honestly do, it says so about rather than faking. There is no
 * per-app grant model here: adb holds shell-level authority over the whole
 * device, disclosed in the capability report and the session notice rather than
 * dressed up as something narrower.
 */
class AdbLink : Link {
    override val label: String = "adb (device-wide shell authority)"

    /** Centres and bounds of the nodes from the last ui_tree, so taps address by id. */
    private val nodeCenters = HashMap<String, IntArray>()   // id -> [x, y]
    private val nodeBounds = HashMap<String, Bounds>()

    override fun close() {
        // Nothing persistent to close; every call spawns adb afresh.
    }

    override fun send(request: JsonObject, timeoutMs: Long): JsonObject {
        return try {
            when (request.opOr()) {
                "capabilities" -> capabilities()
                "session_begin" -> sessionBegin(request.intOr("ttlSec", 300))
                "session_end" -> ok("foregroundChanged" to JsonPrimitive(false))
                "grants_list" -> ok("grants" to JsonArray(emptyList()))
                "apps_list" -> appsList(request.boolOr("grantedOnly", true))
                "ui_tree" -> uiTree()
                "screenshot" -> screenshot()
                "notifications" -> notifications()
                "tap" -> tap(request)
                "long_press" -> longPress(request)
                "swipe" -> swipe(request)
                "scroll" -> scroll(request)
                "type" -> type(request)
                "key" -> key(request.str("key") ?: "")
                "launch" -> launch(request.str("pkg") ?: "")
                "shell" -> shell(request.str("cmd") ?: "")
                "clipboard_get", "clipboard_set" -> err(
                    "clipboard is not available over adb — pair the on-device app for clipboard access",
                )
                else -> err("unsupported op: ${request.opOr()}")
            }
        } catch (e: Exception) {
            val message = e.message ?: e.toString()
            if (message.contains("ENOENT")) {
                err("adb not found on PATH — install platform-tools, or set MCPSERVED_ADB to the adb binary")
            } else {
                err(message)
            }
        }
    }

    // ---- capabilities & session -------------------------------------------

    private fun capabilities(): JsonObject {
        if (!Adb.ready()) {
            return err(
                "no adb device — attach over USB and run `adb devices`, or `adb connect <ip>:5555` for Wi-Fi",
            )
        }

        var root = false
        try {
            val id = Adb.shellText("su -c id", timeoutMs = 5_000)
            root = Regex("uid=0").containsMatchIn(id)
        } catch (_: Exception) {
            root = false
        }

        // adb shell is itself a shell-level backend; report it so the shell tool
        // is offered. "shizuku" here stands for that ADB-level authority.
        val caps = mutableListOf(
            "TREE", "GESTURE", "TEXT_INPUT", "GLOBAL_KEYS", "CAPTURE_SILENT", "NOTIFICATIONS", "SHELL_SHIZUKU",
        )
        if (root) caps.add("SHELL_ROOT")

        return buildJsonObject {
            put("ok", JsonPrimitive(true))
            put("caps", buildJsonArray { caps.forEach { add(it) } })
            put("root", JsonPrimitive(root))
            put("shizuku", JsonPrimitive(true))
            put("a11y", JsonPrimitive(false))
        }
    }

    private fun sessionBegin(ttlSec: Int): JsonObject {
        try {
            Adb.shell("input keyevent 224", timeoutMs = 5_000) // WAKEUP
            Adb.shell("wm dismiss-keyguard", timeoutMs = 5_000)
        } catch (_: Exception) {
            // A device that will not wake is still worth trying to drive.
        }
        return buildJsonObject {
            put("ok", JsonPrimitive(true))
            put("sessionId", JsonPrimitive("adb"))
            put("expiresAtEpochMs", JsonPrimitive(System.currentTimeMillis() + ttlSec * 1000L))
            put("foregroundChanged", JsonPrimitive(false))
        }
    }

    // ---- reading ----------------------------------------------------------

    private fun appsList(grantedOnly: Boolean): JsonObject {
        val flag = if (grantedOnly) " -3" else ""
        val out = Adb.shellText("pm list packages$flag")
        val apps = out.split("\n").map { it.trim() }
            .filter { it.startsWith("package:") }
            .map { it.removePrefix("package:") }
            .filter { it.isNotEmpty() }
            .map { pkg ->
                buildJsonObject {
                    put("pkg", JsonPrimitive(pkg))
                    put("label", JsonPrimitive(pkg))
                    put("granted", JsonPrimitive(true))
                }
            }
        return ok("apps" to JsonArray(apps))
    }

    private fun uiTree(): JsonObject {
        var dumpPath = "/sdcard/window_dump.xml"
        try {
            val out = Adb.shellText("uiautomator dump", timeoutMs = 15_000)
            Regex("""dumped to:\s*(\S+)""").find(out)?.let { dumpPath = it.groupValues[1] }
        } catch (e: Exception) {
            return err("uiautomator dump failed: ${e.message}")
        }

        val xmlRes = Adb.shell("cat ${Adb.shellQuote(dumpPath)}", binary = true, timeoutMs = 15_000)
        if (xmlRes.code != 0) return err("could not read the ui dump")
        val xml = String(xmlRes.stdout, Charsets.UTF_8)

        val parsed = parseUiAutomator(xml)
        nodeCenters.clear()
        nodeBounds.clear()
        for (n in parsed.nodes) {
            nodeCenters[n.id] = intArrayOf((n.bounds.l + n.bounds.r) / 2, (n.bounds.t + n.bounds.b) / 2)
            nodeBounds[n.id] = n.bounds
        }

        return buildJsonObject {
            put("ok", JsonPrimitive(true))
            put("pkg", JsonPrimitive(parsed.pkg))
            put("activity", JsonPrimitive(null as String?))
            put("nodes", JsonArray(parsed.nodes.map { it.toJson() }))
            put("pruned", JsonPrimitive(parsed.pruned))
            put("foregroundChanged", JsonPrimitive(false))
        }
    }

    private fun screenshot(): JsonObject {
        val res = Adb.shell("screencap -p", binary = true, timeoutMs = 20_000)
        if (res.code != 0 || res.stdout.isEmpty()) return err("screencap failed")
        val png = res.stdout
        val (w, h) = pngSize(png)
        // maxPx is not honoured: adb has no image scaler on the host side. The raw
        // frame is returned; the tool description already steers toward ui_tree.
        return buildJsonObject {
            put("ok", JsonPrimitive(true))
            put("mime", JsonPrimitive("image/png"))
            put("b64", JsonPrimitive(Base64.getEncoder().encodeToString(png)))
            put("w", JsonPrimitive(w))
            put("h", JsonPrimitive(h))
            put("foregroundChanged", JsonPrimitive(false))
        }
    }

    private fun notifications(): JsonObject {
        val out = try {
            Adb.shellText("dumpsys notification --noredact", timeoutMs = 10_000)
        } catch (_: Exception) {
            return ok("items" to JsonArray(emptyList()))
        }

        val items = mutableListOf<JsonObject>()
        val blocks = out.split("NotificationRecord(").drop(1)
        for (block in blocks) {
            val pkg = Regex("""pkg=(\S+)""").find(block)?.groupValues?.get(1) ?: ""
            if (pkg.isEmpty()) continue
            val title = Regex("""android\.title=String \(([^)]*)\)""").find(block)?.groupValues?.get(1)
            val text = Regex("""android\.text=String \(([^)]*)\)""").find(block)?.groupValues?.get(1)
            items.add(buildJsonObject {
                put("pkg", JsonPrimitive(pkg))
                put("key", JsonPrimitive(pkg))
                put("title", JsonPrimitive(title))
                put("text", JsonPrimitive(text))
                put("postedAtEpochMs", JsonPrimitive(0))
            })
        }
        return ok("items" to JsonArray(items))
    }

    // ---- acting -----------------------------------------------------------

    private fun resolvePoint(req: JsonObject): IntArray? {
        req.str("nodeId")?.let { return nodeCenters[it] }
        val x = req.str("x")?.toIntOrNull()
        val y = req.str("y")?.toIntOrNull()
        return if (x != null && y != null) intArrayOf(x, y) else null
    }

    private fun tap(req: JsonObject): JsonObject {
        val p = resolvePoint(req) ?: return err("no node id (call ui_tree first) or x,y given")
        Adb.shellText("input tap ${p[0]} ${p[1]}")
        return ok("foregroundChanged" to JsonPrimitive(false))
    }

    private fun longPress(req: JsonObject): JsonObject {
        val p = resolvePoint(req) ?: return err("no node id or x,y given")
        val ms = req.intOr("ms", 500)
        Adb.shellText("input swipe ${p[0]} ${p[1]} ${p[0]} ${p[1]} $ms")
        return ok("foregroundChanged" to JsonPrimitive(false))
    }

    private fun swipe(req: JsonObject): JsonObject {
        val ms = req.intOr("ms", 300)
        val x1 = req.intOr("x1", 0); val y1 = req.intOr("y1", 0)
        val x2 = req.intOr("x2", 0); val y2 = req.intOr("y2", 0)
        Adb.shellText("input swipe $x1 $y1 $x2 $y2 $ms")
        return ok("foregroundChanged" to JsonPrimitive(false))
    }

    private fun scroll(req: JsonObject): JsonObject {
        val b = req.str("nodeId")?.let { nodeBounds[it] } ?: return err("unknown node id — call ui_tree first")
        val cx = (b.l + b.r) / 2
        val cy = (b.t + b.b) / 2
        val w = b.r - b.l
        val h = b.b - b.t
        var x1 = cx; var y1 = cy; var x2 = cx; var y2 = cy
        when (req.str("dir")) {
            "DOWN" -> { y1 = b.t + (h * 0.7).toInt(); y2 = b.t + (h * 0.3).toInt() }
            "UP" -> { y1 = b.t + (h * 0.3).toInt(); y2 = b.t + (h * 0.7).toInt() }
            "RIGHT" -> { x1 = b.l + (w * 0.7).toInt(); x2 = b.l + (w * 0.3).toInt() }
            "LEFT" -> { x1 = b.l + (w * 0.3).toInt(); x2 = b.l + (w * 0.7).toInt() }
            else -> return err("unknown direction: ${req.str("dir")}")
        }
        Adb.shellText("input swipe $x1 $y1 $x2 $y2 300")
        return ok("foregroundChanged" to JsonPrimitive(false))
    }

    private fun type(req: JsonObject): JsonObject {
        req.str("nodeId")?.let { id -> nodeCenters[id]?.let { Adb.shellText("input tap ${it[0]} ${it[1]}") } }
        // `input text` takes spaces as %s and cannot express newlines or most
        // non-ASCII. Good enough for field entry; the app path handles the rest.
        val encoded = (req.str("text") ?: "").replace(" ", "%s")
        Adb.shellText("input text ${Adb.shellQuote(encoded)}")
        return ok("foregroundChanged" to JsonPrimitive(false))
    }

    private fun key(key: String): JsonObject {
        if (key == "NOTIFICATIONS") {
            Adb.shellText("cmd statusbar expand-notifications")
            return ok("foregroundChanged" to JsonPrimitive(false))
        }
        val code = mapOf(
            "BACK" to "4", "HOME" to "3", "RECENTS" to "187", "ENTER" to "66", "DELETE" to "67",
        )[key] ?: return err("unknown key: $key")
        Adb.shellText("input keyevent $code")
        return ok("foregroundChanged" to JsonPrimitive(false))
    }

    private fun launch(pkg: String): JsonObject {
        val r = Adb.shell(
            "monkey -p ${Adb.shellQuote(pkg)} -c android.intent.category.LAUNCHER 1",
            timeoutMs = 10_000,
        )
        if (r.code != 0) return err("could not launch $pkg")
        return ok("foregroundChanged" to JsonPrimitive(true))
    }

    private fun shell(cmd: String): JsonObject {
        val r = Adb.shell(cmd, timeoutMs = 30_000)
        val text = (String(r.stdout, Charsets.UTF_8) + if (r.stderr.isNotEmpty()) "\n${r.stderr}" else "").trim()
        return buildJsonObject {
            put("ok", JsonPrimitive(r.code == 0))
            put("error", JsonPrimitive(if (r.code == 0) null else text.ifEmpty { "exited ${r.code}" }))
            put("text", JsonPrimitive(text))
        }
    }
}

// ---- parsing helpers ------------------------------------------------------

private data class Bounds(val l: Int, val t: Int, val r: Int, val b: Int)

private class AdbNode(
    val id: String,
    val cls: String,
    val bounds: Bounds,
    val depth: Int,
    val text: String?,
    val desc: String?,
    val clickable: Boolean,
    val editable: Boolean,
    val scrollable: Boolean,
    val checked: Boolean?,
    val enabled: Boolean,
) {
    fun toJson(): JsonObject = buildJsonObject {
        put("id", JsonPrimitive(id))
        put("cls", JsonPrimitive(cls))
        put("bounds", buildJsonObject {
            put("l", JsonPrimitive(bounds.l)); put("t", JsonPrimitive(bounds.t))
            put("r", JsonPrimitive(bounds.r)); put("b", JsonPrimitive(bounds.b))
        })
        put("depth", JsonPrimitive(depth))
        put("text", JsonPrimitive(text))
        put("desc", JsonPrimitive(desc))
        put("clickable", JsonPrimitive(clickable))
        put("editable", JsonPrimitive(editable))
        put("scrollable", JsonPrimitive(scrollable))
        put("checked", JsonPrimitive(checked))
        put("enabled", JsonPrimitive(enabled))
    }
}

private class ParsedTree(val pkg: String, val nodes: List<AdbNode>, val pruned: Int)

private fun decodeEntities(s: String): String = s
    .replace("&amp;", "&").replace("&lt;", "<").replace("&gt;", ">")
    .replace("&quot;", "\"").replace("&#39;", "'").replace("&apos;", "'")

private fun parseBounds(s: String?): Bounds? {
    val m = s?.let { Regex("""\[(-?\d+),(-?\d+)]\[(-?\d+),(-?\d+)]""").find(it) } ?: return null
    val (a, b, c, d) = m.destructured
    return Bounds(a.toInt(), b.toInt(), c.toInt(), d.toInt())
}

/**
 * Parses a uiautomator hierarchy dump into the flat node list the tools expect.
 *
 * Pure layout containers — nodes with no text, no content description, and no
 * interactive flag — are dropped and counted, matching what the on-device Pruner
 * does, so the tree the model reads is interaction, not scaffolding.
 */
private fun parseUiAutomator(xml: String): ParsedTree {
    val nodes = mutableListOf<AdbNode>()
    var pruned = 0
    var pkg = ""
    var depth = 0
    var seq = 0

    val tag = Regex("""<node\b([^>]*?)(/?)>|</node>""")
    val attrRe = Regex("""([\w-]+)="([^"]*)"""")

    for (m in tag.findAll(xml)) {
        if (m.value == "</node>") {
            depth = maxOf(0, depth - 1)
            continue
        }
        val attrsRaw = m.groupValues[1]
        val selfClosing = m.groupValues[2] == "/"
        val attrs = HashMap<String, String>()
        for (a in attrRe.findAll(attrsRaw)) attrs[a.groupValues[1]] = decodeEntities(a.groupValues[2])

        val bounds = parseBounds(attrs["bounds"])
        val thisDepth = depth
        if (!selfClosing) depth += 1
        if (bounds == null) continue
        if (pkg.isEmpty() && attrs["package"] != null) pkg = attrs["package"]!!

        val text = attrs["text"]?.ifEmpty { null }
        val desc = attrs["content-desc"]?.ifEmpty { null }
        val clickable = attrs["clickable"] == "true"
        val longClickable = attrs["long-clickable"] == "true"
        val scrollable = attrs["scrollable"] == "true"
        val checkable = attrs["checkable"] == "true"
        val rawCls = attrs["class"] ?: ""
        val cls = rawCls.substringAfterLast(".").ifEmpty { rawCls.ifEmpty { "View" } }
        val editable = Regex("EditText", RegexOption.IGNORE_CASE).containsMatchIn(rawCls)

        val interactive = clickable || longClickable || scrollable || checkable || editable
        if (!interactive && text == null && desc == null) { pruned += 1; continue }
        if (bounds.r - bounds.l <= 0 || bounds.b - bounds.t <= 0) { pruned += 1; continue }

        val rid = attrs["resource-id"]
        val idBase = if (!rid.isNullOrEmpty()) rid.substringAfterLast("/").ifEmpty { rid } else cls.lowercase()
        val id = "$idBase#${seq++}"

        nodes.add(
            AdbNode(
                id = id, cls = cls, bounds = bounds, depth = thisDepth,
                text = text, desc = desc, clickable = clickable, editable = editable,
                scrollable = scrollable,
                checked = if (checkable) attrs["checked"] == "true" else null,
                enabled = attrs["enabled"] != "false",
            ),
        )
    }
    return ParsedTree(pkg, nodes, pruned)
}

/** Reads width and height from a PNG's IHDR chunk. */
private fun pngSize(png: ByteArray): Pair<Int, Int> {
    if (png.size >= 24 && String(png, 12, 4, Charsets.US_ASCII) == "IHDR") {
        fun u32(off: Int): Int =
            ((png[off].toInt() and 0xff) shl 24) or ((png[off + 1].toInt() and 0xff) shl 16) or
                ((png[off + 2].toInt() and 0xff) shl 8) or (png[off + 3].toInt() and 0xff)
        return u32(16) to u32(20)
    }
    return 0 to 0
}
